File: solutions/R/applied_bayesian_modeling_ex2.R
# Chapter 14, Exercise 2: Hierarchical Model for Multi-Site Drug Trial
# 12 hospitals, LDL cholesterol change, new statin vs standard care

library(brms)
library(tidyverse)

# ---- (a) Simulate data ----
set.seed(789)
n_hospitals <- 12
patients_per_hospital <- c(20, 25, 30, 40, 50, 60, 70, 80, 100, 120, 150, 200)
true_grand_effect <- -25  # mg/dL reduction in LDL
tau_true <- 5             # between-hospital SD

hospital_effects <- rnorm(n_hospitals, 0, tau_true)

trial_data <- map2_dfr(1:n_hospitals, patients_per_hospital, function(j, nj) {
  tibble(
    hospital = factor(j),
    treatment = rep(0:1, length.out = nj),
    ldl_change = (true_grand_effect + hospital_effects[j]) * treatment +
      rnorm(nj, 0, 20)  # residual SD = 20 mg/dL
  )
})

cat("Data summary:\n")
cat("  Total patients:", nrow(trial_data), "\n")
cat("  Hospitals:", n_hospitals, "\n")
cat("  Patients per hospital:", paste(patients_per_hospital, collapse = ", "), "\n")
cat("  True grand effect:", true_grand_effect, "mg/dL\n")
cat("  True between-hospital SD:", tau_true, "mg/dL\n")

# ---- (b) Fit Bayesian hierarchical model ----
fit_hier <- brm(
  ldl_change ~ treatment + (treatment | hospital),
  data = trial_data,
  prior = c(
    prior(normal(0, 30), class = "b"),
    prior(normal(0, 30), class = "Intercept"),
    prior(cauchy(0, 5), class = "sd"),
    prior(exponential(0.05), class = "sigma")
  ),
  chains = 4,
  iter = 2000,
  warmup = 1000,
  seed = 42,
  silent = 2,
  refresh = 0,
  control = list(adapt_delta = 0.95)
)

cat("\nHierarchical model summary:\n")
summary(fit_hier)

# Grand mean treatment effect
grand_mean <- fixef(fit_hier)["treatment", "Estimate"]
cat("\nGrand mean treatment effect:", round(grand_mean, 1), "mg/dL\n")

# ---- (c) Shrinkage plot ----
# Extract hospital-specific treatment effects (partial pooling)
ranefs <- ranef(fit_hier)$hospital[, , "treatment"]
partial_pool <- grand_mean + ranefs[, "Estimate"]

# No-pooling estimates (separate OLS per hospital)
no_pool <- trial_data %>%
  group_by(hospital) %>%
  summarise(
    no_pool_est = coef(lm(ldl_change ~ treatment))[2],
    n = n(),
    .groups = "drop"
  ) %>%
  mutate(
    hospital_num = as.numeric(hospital),
    partial_pool_est = partial_pool
  )

# Plot shrinkage
ggplot(no_pool, aes(y = reorder(hospital, n))) +
  geom_point(aes(x = no_pool_est, colour = "No pooling"), size = 3) +
  geom_point(aes(x = partial_pool_est, colour = "Partial pooling"), size = 3) +
  geom_vline(xintercept = grand_mean, linetype = "dashed", colour = "grey40") +
  geom_segment(aes(x = no_pool_est, xend = partial_pool_est,
                   yend = reorder(hospital, n)),
               arrow = arrow(length = unit(0.15, "cm")),
               colour = "grey60") +
  annotate("text", x = grand_mean + 1, y = 12.5,
           label = paste0("Grand mean = ", round(grand_mean, 1)),
           hjust = 0, size = 3.5) +
  scale_colour_manual(values = c("No pooling" = "steelblue",
                                  "Partial pooling" = "firebrick")) +
  labs(x = "Treatment Effect (mg/dL change in LDL)",
       y = "Hospital (ordered by sample size)",
       colour = "Estimate Type",
       title = "Shrinkage in Hierarchical Model",
       subtitle = "Arrows show partial pooling pulling estimates toward the grand mean") +
  theme_minimal(base_size = 13) +
  theme(legend.position = "bottom")

# ---- (d) Compare hierarchical to separate OLS regressions ----
cat("\n=== Part (d): Comparison ===\n\n")
cat("Hospital  |  N   | No-Pooling | Partial Pooling | Shrinkage\n")
cat("----------|------|------------|-----------------|----------\n")

for (i in 1:nrow(no_pool)) {
  shrinkage <- abs(no_pool$no_pool_est[i] - no_pool$partial_pool_est[i])
  cat(sprintf("   %2d     | %3d  |   %6.1f   |     %6.1f      |  %5.1f\n",
              no_pool$hospital_num[i], no_pool$n[i],
              no_pool$no_pool_est[i], no_pool$partial_pool_est[i], shrinkage))
}

cat("\nThe hierarchical model is MORE APPROPRIATE because:\n")
cat("1. Small hospitals (n=20-30) have noisy OLS estimates that are\n")
cat("   shrunk toward the grand mean, reducing estimation error.\n")
cat("2. Large hospitals (n=150-200) retain their individual estimates\n")
cat("   since their data are informative enough.\n")
cat("3. The between-hospital SD (tau) is estimated from the data,\n")
cat("   quantifying the degree of heterogeneity across sites.\n")
cat("4. Hospital-by-hospital OLS ignores the shared structure --\n")
cat("   all hospitals are studying the same drug. The hierarchical\n")
cat("   model borrows strength across sites while allowing for\n")
cat("   genuine between-site variation.\n")

File: solutions/python/applied_bayesian_modeling_ex2.py
# Chapter 14, Exercise 2: Hierarchical Model for Multi-Site Drug Trial
# 12 hospitals, LDL cholesterol change, new statin vs standard care

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from sklearn.linear_model import LinearRegression

# ---- (a) Simulate data ----
np.random.seed(789)
n_hospitals = 12
patients = [20, 25, 30, 40, 50, 60, 70, 80, 100, 120, 150, 200]
grand_effect = -25  # mg/dL
tau = 5             # between-hospital SD

hospital_effects = np.random.normal(0, tau, n_hospitals)

rows = []
for j in range(n_hospitals):
    nj = patients[j]
    trt = np.tile([0, 1], nj // 2 + 1)[:nj]
    ldl_change = ((grand_effect + hospital_effects[j]) * trt +
                  np.random.normal(0, 20, nj))
    for i in range(nj):
        rows.append({
            'hospital': j,
            'treatment': trt[i],
            'ldl_change': ldl_change[i]
        })

trial_data = pd.DataFrame(rows)

print("Data summary:")
print(f"  Total patients: {len(trial_data)}")
print(f"  Hospitals: {n_hospitals}")
print(f"  True grand effect: {grand_effect} mg/dL")
print(f"  True between-hospital SD: {tau} mg/dL")

# ---- No-pooling estimates (OLS per hospital) ----
no_pool = []
for j in range(n_hospitals):
    df_j = trial_data[trial_data['hospital'] == j]
    trt_mean = df_j[df_j['treatment'] == 1]['ldl_change'].mean()
    ctl_mean = df_j[df_j['treatment'] == 0]['ldl_change'].mean()
    no_pool.append(trt_mean - ctl_mean)

no_pool = np.array(no_pool)

# ---- (b) & (c) Partial pooling approximation ----
# For a full Bayesian fit, use PyMC. Here we approximate shrinkage
# to demonstrate the concept without requiring MCMC sampling.

# Estimate within-hospital variance from data
within_var = []
for j in range(n_hospitals):
    df_j = trial_data[trial_data['hospital'] == j]
    trt_vals = df_j[df_j['treatment'] == 1]['ldl_change'].values
    ctl_vals = df_j[df_j['treatment'] == 0]['ldl_change'].values
    # Variance of the treatment effect estimate
    se_j = np.sqrt(np.var(trt_vals, ddof=1) / len(trt_vals) +
                   np.var(ctl_vals, ddof=1) / len(ctl_vals))
    within_var.append(se_j**2)

within_var = np.array(within_var)

# Estimate between-hospital variance using method of moments
grand_mean_est = np.mean(no_pool)
tau_est_sq = max(0, np.var(no_pool, ddof=1) - np.mean(within_var))
tau_est = np.sqrt(tau_est_sq)

# Shrinkage factor: B_j = within_var_j / (within_var_j + tau^2)
shrinkage = within_var / (within_var + tau_est_sq)
partial_pool = grand_mean_est + (1 - shrinkage) * (no_pool - grand_mean_est)

print(f"\nEstimated grand mean effect: {grand_mean_est:.1f} mg/dL")
print(f"Estimated between-hospital SD: {tau_est:.1f} mg/dL")

# ---- (c) Shrinkage plot ----
fig, ax = plt.subplots(figsize=(10, 6))
order = np.argsort(patients)
y_pos = np.arange(n_hospitals)

for i, idx in enumerate(order):
    ax.annotate("", xy=(partial_pool[idx], i), xytext=(no_pool[idx], i),
                arrowprops=dict(arrowstyle="->", color="grey"))

ax.scatter(no_pool[order], y_pos, color="steelblue", s=60, zorder=2,
           label="No pooling (OLS)")
ax.scatter(partial_pool[order], y_pos, color="firebrick", s=60, zorder=2,
           label="Partial pooling")
ax.axvline(grand_mean_est, linestyle="--", color="grey", linewidth=1)
ax.text(grand_mean_est + 0.5, n_hospitals - 0.5,
        f"Grand mean = {grand_mean_est:.1f}", fontsize=9)

ax.set_yticks(y_pos)
ax.set_yticklabels([f"Hospital {order[i]+1}\n(n={patients[order[i]]})"
                     for i in range(n_hospitals)], fontsize=9)
ax.set_xlabel("Treatment Effect (mg/dL change in LDL)")
ax.set_title("Shrinkage in Hierarchical Model\n"
             "Arrows show estimates pulled toward the grand mean")
ax.legend()
plt.tight_layout()
plt.savefig("ch14_ex2_shrinkage.png", dpi=150)
plt.show()

# ---- (d) Compare approaches ----
print("\n=== Part (d): Comparison ===\n")
print(f"{'Hospital':>8} | {'N':>4} | {'No-Pool':>10} | {'Partial Pool':>12} | {'Shrinkage':>9}")
print("-" * 55)
for j in range(n_hospitals):
    shrink_amt = abs(no_pool[j] - partial_pool[j])
    print(f"   {j+1:>2}     | {patients[j]:>3}  | {no_pool[j]:>9.1f}  | {partial_pool[j]:>11.1f}  | {shrink_amt:>8.1f}")

print("\nThe hierarchical model is MORE APPROPRIATE because:")
print("1. Small hospitals have noisy OLS estimates that are shrunk")
print("   toward the grand mean, reducing estimation error.")
print("2. Large hospitals retain their individual estimates since")
print("   their data are informative enough.")
print("3. Between-hospital SD (tau) is estimated from data,")
print("   quantifying heterogeneity across sites.")
print("4. Hospital-by-hospital OLS ignores shared structure --")
print("   all hospitals study the same drug. The hierarchical model")
print("   borrows strength across sites while allowing genuine")
print("   between-site variation.")

# NOTE: For a full Bayesian hierarchical model, use PyMC:
#
# import pymc as pm
# with pm.Model() as hier_model:
#     mu_trt = pm.Normal("mu_trt", mu=0, sigma=30)
#     tau = pm.HalfCauchy("tau", beta=5)
#     trt_j = pm.Normal("trt_j", mu=mu_trt, sigma=tau, shape=n_hospitals)
#     sigma = pm.Exponential("sigma", lam=0.05)
#     mu = trt_j[hospital_idx] * treatment
#     y = pm.Normal("y", mu=mu, sigma=sigma, observed=ldl_change)
#     trace = pm.sample(1000, tune=1000, chains=4, random_seed=42)
